documentapi/src/main/java/com/yahoo/documentapi/messagebus/protocol/RoutableFactories80.java

```
@@ -0,0 +1,931 @@
+                    protoStream.flush();
+                }
+            } catch (IOException | UnsupportedOperationException e) {
+                return false;
+            }
+            return true;
+        }
+
+        @Override
+        public Routable decode(DocumentDeserializer in) {
+            return decoderFn.apply(in);
+        }
+    }
+
+    private static class ProtobufCodecBuilder<DocApiT extends Routable, ProtoT extends AbstractMessage> {
+
+        private final Class<DocApiT>                    apiClass;
+        private final Class<ProtoT>                     protoClass;
+        private Function<DocApiT, ProtoT>               encoderFn;
+        private Function<DocumentDeserializer, DocApiT> decoderFn;
+
+        ProtobufCodecBuilder(Class<DocApiT> apiClass, Class<ProtoT> protoClass) {
+            this.apiClass = apiClass;
+            this.protoClass = protoClass;
+        }
+
+        static <DocApiT extends Routable, ProtoT extends AbstractMessage> ProtobufCodecBuilder<DocApiT, ProtoT>
+        of(Class<DocApiT> apiClass, Class<ProtoT> protoClass) {
+            return new ProtobufCodecBuilder<>(apiClass, protoClass);
+        }
+
+        ProtobufCodecBuilder<DocApiT, ProtoT> encoder(Function<DocApiT, ProtoT> fn) {
+            if (encoderFn != null) {
+                throw new IllegalArgumentException("Encoder already set");
+            }
+            encoderFn = fn;
+            return this;
+        }
+
+        ProtobufCodecBuilder<DocApiT, ProtoT> decoder(ProtoT messagePrototype, Function<ProtoT, DocApiT> fn) {
+            if (decoderFn != null) {
+                throw new IllegalArgumentException("Decoder already set");
+            }
+            decoderFn = (buf) -> {
+                try {
+                    var protoObj = messagePrototype.getParserForType().parseFrom(buf.getBuf().getByteBuffer());
+                    return fn.apply(protoClass.cast(protoObj));
+                } catch (IOException e) {
```

Are we guaranteed the `IOException` is never thrown?

---

Yes, but with no trace of what went wrong?

---

Good point. I see there is a higher-level `decode` method which does at least print out the exception message as a warning. Will runtime-wrap and re-throw instead of just returning null

---

I think that's a good idea.

---

Now has explicit catching and logging for both encoding and decoding